ssl/statem/statem_srvr.c

```diff
@@ -3679,12 +3691,25 @@ MSG_PROCESS_RETURN tls_process_client_certificate(SSL *s, PACKET *pkt)
     sk = NULL;
 
     /* Save the current hash state for when we receive the CertificateVerify */
-    if (SSL_IS_TLS13(s)
-            && !ssl_handshake_hash(s, s->cert_verify_hash,
-                                   sizeof(s->cert_verify_hash),
-                                   &s->cert_verify_hash_len)) {
-        /* SSLfatal() already called */
-        goto err;
+    if (SSL_IS_TLS13(s)) {
+        size_t j;
+
+        if (!ssl_handshake_hash(s, s->cert_verify_hash,
+                                sizeof(s->cert_verify_hash),
+                                &s->cert_verify_hash_len)) {
+            /* SSLfatal() already called */
+            goto err;
+        }
+
+        /* Invalidate old tickets */
+        for (j = 0; j < s->sent_tickets; j++) {
+            SSL_CTX_remove_session(s->session_ctx, s->tickets[j]);
+            SSL_SESSION_free(s->tickets[j]);
+            s->tickets[j] = NULL;
+        }
```

Comment:
I'm not convinced that this invalidation of old tickets is either necessary or effective. Not all the tickets issued as a result of some previous full handshake and its resumptions will have been issued by *this* connection. What purpose is served by explicitly holding on to the ones issued before PHA?

Reply:
TBH, I'm not really sure why I thought it was necessary when I wrote this code. Looking back at it now, it seems over the top, so I have removed it.